src/snowflake/snowpark/udf.py

```diff
@@ -897,3 +915,115 @@ def _do_register_udf(
+        from_pandas_udf_function: bool = False,
+        skip_upload_on_content_match: bool = False,
+    ) -> UserDefinedFunction:
+
+        (
+            udf_name,
+            is_pandas_udf,
+            is_dataframe_input,
+            return_type,
+            input_types,
+        ) = process_registration_inputs(
+            session=self._session,
+            object_type=callableProperties.object_type,
+            func=callableProperties.func,
+            return_type=callableProperties.raw_return_type,
+            input_types=callableProperties.raw_input_types,
+            name=callableProperties.raw_name,
+        )
+
+        callableProperties.setValidatedObjectName(udf_name)
+        callableProperties.setValidatedReturnType(return_type)
+        callableProperties.setValidatedInputTypes(input_types)
+
+        arg_names = [f"arg{i + 1}" for i in range(len(input_types))]
+        callableProperties.setResolvedInputArgs(
+            [UDFColumn(dt, arg_name) for dt, arg_name in zip(input_types, arg_names)]
+        )
+
+        # allow registering pandas UDF from udf(),
+        # but not allow registering non-pandas UDF from pandas_udf()
+        if from_pandas_udf_function and not is_pandas_udf:
+            raise ValueError(
+                "You cannot create a non-vectorized UDF using pandas_udf(). "
+                "Use udf() instead."
+            )
+
+        resolved_packages = resolve_packages(
+            session=self._session,
+            packages=callableProperties.raw_packages,
+            is_pandas_udf=is_pandas_udf,
+            statement_params=statement_params,
+        )
+        callableProperties.setResolvedPackages(resolved_packages)
+
+        (
+            handler,
+            inline_code,
+            all_imports,
+            upload_file_stage_location,
+            custom_python_runtime_version_allowed,
+        ) = resolve_imports(
+            session=self._session,
+            callableProperties=callableProperties,
+            stage_location=stage_location,
+            arg_names=arg_names,
+            statement_params=statement_params,
+            is_dataframe_input=is_dataframe_input,
+            skip_upload_on_content_match=skip_upload_on_content_match,
+            is_pandas_udf=is_pandas_udf,
+        )
+
+        callableProperties.setResolvedHandler(handler)
+        callableProperties.setResolvedInlineCode(inline_code)
+        callableProperties.setResolvedImports(all_imports)
+
+        if not custom_python_runtime_version_allowed:
+            check_python_runtime_version(
+                self._session._runtime_version_from_requirement
+            )
+
+        runtime_version = (
+            f"{sys.version_info[0]}.{sys.version_info[1]}"
+            if not self._session._runtime_version_from_requirement
+            else self._session._runtime_version_from_requirement
+        )
+        callableProperties.setResolvedRuntimeVersion(runtime_version)
+
+        raised = False
```

Comment:
At this point, the hook introduced by the CLI will be invoked to send the `callableProperties` data back to the CLI, instead of invoking `create_python_udf_or_sp`.